[DOCSP-24548] Update fundamentals page (#33)
mongodb/docs-csharp: source/fundamentals/crud/write-operations/update.txt

.. _csharp-update-guide:

================

> Update Documents

Author: Talked to Jordan about this. We'd both prefer to keep them separate.

================

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

In this guide, you'll learn how to edit existing documents or upsert new documents 
in your MongoDB collections using update operations.

Sample Data
~~~~~~~~~~~

The examples in this guide use the ``restaurants`` collection
from the ``sample_restaurants`` database. The documents in this
collection use the following ``Restaurant`` class as a model:

.. literalinclude:: /includes/fundamentals/code-examples/crud/update/Restaurant.cs
   :language: csharp
   :dedent:
   :start-after: start-model
   :end-before: end-model

This collection is from the :atlas:`sample datasets </sample-data>` provided
by Atlas. See the :ref:`<csharp-quickstart>` to learn how to create a free MongoDB cluster
and load this sample data.

Update Operations
-----------------

You can perform update operations in MongoDB with the following methods:

> - ``UpdateOne()``, which updates *the first document* that matches the query filter.

Author: The ordering gets a little funny here because the very next section describes what a query filter is, and it feels more natural to do so in the context of the parameters rather than the context of the method itself. One option I see is to reword the sentence you called out to something like "search" instead of "query filter."  What do you think?

> - ``UpdateOne()``, which updates *the first document* that matches the query filter.

Reviewer: maybe search criteria?

> - ``UpdateMany()``, which updates *all documents* that match the query filter.

Author: See above

Required Parameters
~~~~~~~~~~~~~~~~~~~

> Update operations require two parameters: 

Reviewer: ```suggestion
Each update method requires the following parameters:
```

- A **query filter** document, which determines which records to update. See the 
  :manual:`MongoDB server manual </core/document/#query-filter-documents>` for 
  more information about query filters.
- An **update** document, which specifies the **update operator** (the kind of update to
  perform) and the fields and values that should change. See the 

>   :manual:`MongoDB server manual</reference/operator/update-field/>` for a complete 

Reviewer: ```suggestion
  :manual:`Field Update Operators Manual page</reference/operator/update-field/>` for a complete 
```

  list of update operators and their usage. 

> The {+driver-short+} includes a ``Builders`` class that simplifies the creation of both

Reviewer: ```suggestion
The {+driver-short+} provides a ``Builders`` class that simplifies the creation of both
```

> query filters and update documents. The following code sample uses ``Builders`` to create
> a query filter that searches for restaurants with a ``borough`` field value of "Manhattan". 
> Then it uses ``Builders`` to create an update document that sets the value of the 
> ``borough`` field of these restaurants to "Manhattan (north)":

Reviewer: ```suggestion
query filters and update documents. The following code sample uses ``Builders`` to create:
- A filter that matches documents with the ``borough`` field value "Manhattan"
- An update document that sets the value of the ``borough`` field to "Manhattan (north)"
```

Author: Changed to a list, but reworded your suggestion to avoid introducing the items with a sentence fragment.

.. literalinclude:: /includes/fundamentals/code-examples/crud/update/Update.cs
   :language: csharp
   :dedent:
   :start-after: start-builders

>    :end-before: end-builders

Author: I think my reasoning was that the user could see builders in isolation and with explanation, then see them put to use in the full example later.

>    :end-before: end-builders

Author: let's zoom about this when you're ready. i'm not sure i follow

.. tabs::

   .. tab:: Asynchronous
      :tabid: update-one-async

      .. code-block:: csharp
         :copyable: true

         var result = await _restaurantsCollection.UpdateOneAsync(filter, update);

   .. tab:: Synchronous
      :tabid: update-one-sync

      .. code-block:: csharp
         :copyable: true

         var result = _restaurantsCollection.UpdateOne(filter, update);

Update Many Documents
~~~~~~~~~~~~~~~~~~~~~

The following code shows how to use the asynchronous
``UpdateManyAsync()`` method or the synchronous ``UpdateMany()`` method to
update all matched documents.

.. tabs::

   .. tab:: Asynchronous
      :tabid: update-many-async

      .. code-block:: csharp
         :copyable: true

         var result = await _restaurantsCollection.UpdateManyAsync(filter, update);

   .. tab:: Synchronous
      :tabid: update-many-sync

      .. code-block:: csharp
         :copyable: true

         var result = _restaurantsCollection.UpdateMany(filter, update);

.. tip::

   Find runnable examples that use these methods under :ref:`Additional
   Information <csharp-update-info>`.

Customize the Update Operation
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> If you need to customize your update operation, create an ``UpdateOptions`` object 
> and pass it to ``UpdateOne()`` or ``UpdateMany()``. The ``UpdateOptions`` type lets 
> you configure the following properties:

Reviewer: ```suggestion
Both methods optionally take an ``UpdateOptions`` type as an additional parameter,
which represents options you can use to configure the delete operation.
If you don't specify any ``UpdateOptions`` properties, the driver does
not customize the update operation.

The ``UpdateOptions`` type allows you to configure options with the
following properties:
```

     - Description

   * - ``ArrayFilters``
     - | Specifies which array elements to modify for an update operation on an array field.
         See :manual:`the MongoDB server manual</reference/command/update/#update-elements-match-arrayfilters-criteria>`
         for more information.

   * - ``BypassDocumentValidation``
     - | Specifies whether the update operation bypasses document validation. This lets you 
         update documents that don't meet the schema validation requirements, if any 
         exist. See :manual:`the MongoDB server manual</core/schema-validation/#schema-validation>`
         for more information on schema validation.

   * - ``Collation``
     - | Specifies the kind of language collation to use when sorting
         results. See :manual:`the MongoDB server manual</reference/collation/#std-label-collation>`
         for more information on collation.

   * - ``Comment``
     - | Gets or sets the user-provided comment for the operation. 
         See :manual:`the MongoDB server manual</reference/command/update/#command-fields>`
         for more information.

   * - ``Hint``
     - | Gets or sets the index to use to scan for documents. 

>          See :manual:`the MongoDB server manual<reference/command/update/#std-label-update-command-hint>`

Author: fixed

         for more information.

   * - ``IsUpsert``
     - | Specifies whether the update operation performs an upsert operation if no 
         documents match the query filter. 
         See :manual:`the MongoDB server manual </reference/command/update/#std-label-update-command-upsert>`

>          for more information.

Author: I added it to the example. LMK if that's what you had in mind.

   * - ``Let``
     - | Gets or sets the let document. 
         See :manual:`the MongoDB server manual </reference/command/update/#std-label-update-command-upsert>`
         for more information.

.. tip::

    The ``UpdateOptions`` parameter is optional. Use it only if you must change the

>     default update behavior.

Author: Removed

Return Value
~~~~~~~~~~~~

The ``UpdateOne()`` and ``UpdateMany()`` methods each return an ``UpdateResult`` 

> object. The ``UpdateResult`` type contains the following properties:

Reviewer: ```suggestion
type. The ``UpdateResult`` type contains the following properties:
```

Author: I thought about this one while writing. Is it correct to say it "returns a type"? Technically speaking, it returns an object _of_ a particular type. I'd prefer to keep this the way it is unless you feel strongly.

Author: yeah, it's a funny distinction. the UpdateResult class/type contains all the properties in that list (i.e., they are defined within the class), but the object you get back from the operation may not (since one or more may not be relevant in a given case).

.. list-table::
   :widths: 30 70
   :header-rows: 1

   * - Property
     - Description

   * - ``IsAcknowledged``
     - | Indicates whether the update operation was acknowledged by MongoDB.

   * - ``IsModifiedCountAvailable``
     - | Indicates whether you can read the count of updated records on the
         ``UpdateResult``.

   * - ``MatchedCount``
     - | The number of documents that matched the query filter, regardless of
         how many were updated.

   * - ``ModifiedCount``
     - | The number of documents updated by the update operation. If an updated
         document is identical to the original, it won't be included in this count.

   * - ``UpsertedId``
     - | The ID of the document that was upserted in the database, if the driver
         performed an upsert.

Example
~~~~~~~

The following code uses the ``UpdateMany()`` method to find all documents where the 
``borough`` field has the value "Manhattan". It then updates the ``borough``

> value in these documents to "Manhattan (north)":

Author: Reworded your suggestion from "and" to ", then" to avoid ambiguity.

.. _csharp-update-guide:

================
Update Documents
================

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

In this guide, you'll learn how to edit existing documents or upsert new documents 
in your MongoDB collections using update operations.

Sample Data
~~~~~~~~~~~

The examples in this guide use the ``restaurants`` collection
from the ``sample_restaurants`` database. The documents in this
collection use the following ``Restaurant`` class as a model:

.. literalinclude:: /includes/fundamentals/code-examples/crud/update/Restaurant.cs
   :language: csharp
   :dedent:
   :start-after: start-model
   :end-before: end-model

This collection is from the :atlas:`sample datasets </sample-data>` provided
by Atlas. See the :ref:`<csharp-quickstart>` to learn how to create a free MongoDB cluster
and load this sample data.

Update Operations
-----------------

You can perform update operations in MongoDB with the following methods:

> - ``UpdateOne()``, which updates *the first document* that matches the query filter.

Reviewer: also I think fragment list items have no punctuation

- ``UpdateMany()``, which updates *all documents* that match the query filter.

Required Parameters
~~~~~~~~~~~~~~~~~~~

Each update method requires the following parameters:

- A **query filter** document, which determines which records to update. See the 
  :manual:`MongoDB server manual </core/document/#query-filter-documents>` for 
  more information about query filters.
- An **update** document, which specifies the **update operator** (the kind of update to
  perform) and the fields and values that should change. See the 
  :manual:`Field Update Operators Manual page</reference/operator/update-field/>` for a complete 
  list of update operators and their usage. 

The {+driver-short+} includes a ``Builders`` class that simplifies the creation of both
query filters and update documents. The following code sample uses ``Builders`` to create
two documents:
- A query filter that searches for restaurants with a ``borough`` field value of "Manhattan" 
- An update document that sets the value of the ``borough`` field of these restaurants 

> to "Manhattan (north)"

Author: reworded slightly
